### ISSUE TITLE
Add synchronization support

### ISSUE THREAD

**Issue by bbannier** [zeek/spicy#23]
(Setting to low priority, we'll do this once the main parsing stuff is in place)

- Spicy:
    - Unit parsing
        - [x] Add "resynchronization mode"

    - Marks
        - [ ] Add type for mark
        - [ ] Add runtime API to insert marks into input
        - [ ] Extend Spicy driver to allow specifying marks

    - Unit properties
        - [ ] `%sync-after`/`%sync-at` at bytes
        - [ ] `%sync-after`/`%sync-at` at regex
        - [ ] `%sync-after`/`%sync-at` at mark
        - [ ] `%sync-after`/`%sync-at` at object

    - Parse attributes
        - [x] `&synchronize`

    - Unit hooks
        - [x] `%synced`

Note that this needs to integrate with "trial mode" (see #24): If there are synchronization hints, entering trial mode enables resynchronization mode as well; see 2.7 in the Spicy paper.  

The Spicy paper may generally be helpful to understand the synchronization strategy.

Old tests in https://github.com/rsmmr/hilti/tree/master/tests/spicy/synchronize.

**Comment by bbannier** [comment]
Closing this as #1054 has landed.
